go/ai/generator.go

```diff
@@ -134,6 +158,86 @@ func (ga *generatorAction) Generate(ctx context.Context, input *GenerateRequest,
 	}
 }
 
+// conformOutput appends a message to the request indicating conformance to the expected schema.
+func conformOutput(input *GenerateRequest) error {
+	if len(input.Output.Schema) > 0 && len(input.Messages) > 0 {
+		jsonBytes, err := json.Marshal(input.Output.Schema)
+		if err != nil {
+			return fmt.Errorf("expected schema is not valid: %w", err)
+		}
+
+		jsonStr := string(jsonBytes)
+		escapedJSON := strconv.Quote(jsonStr)
+		part := &Part{
+			text: fmt.Sprintf("Output should be in JSON format and conform to the following schema:\n\n```%s```", escapedJSON),
+		}
+		input.Messages[len(input.Messages)-1].Content = append(input.Messages[len(input.Messages)-1].Content, part)
+	}
+	return nil
+}
+
+// findValidCandidates finds all candidates that match the expected schema.
+func findValidCandidates(ctx context.Context, resp *GenerateResponse) []*Candidate {
+	candidates := []*Candidate{}
+	for i, c := range resp.Candidates {
+		err := validateCandidate(c, resp.Request.Output)
+		if err == nil {
+			candidates = append(candidates, c)
+		} else {
+			internal.Logger(ctx).Debug("candidate did not match provided output schema", "index", i, "error", err.Error())
+		}
+	}
+	return candidates
+}
+
+// validateCandidate will check a candidate against the expected schema.
+// It will return an error if it does not match, otherwise it will return nil.
+func validateCandidate(candidate *Candidate, outputSchema *GenerateRequestOutput) error {
+	if outputSchema.Format != OutputFormatJSON {
+		return nil
+	}
+
+	text, err := candidate.Text()
+	if err != nil {
+		return err
+	}
+
+	text = stripJsonDelimiters(text)
+
+	var jsonData interface{}
+	err = json.Unmarshal([]byte(text), &jsonData)
+	if err != nil {
+		return fmt.Errorf("candidate did not have valid JSON: %w", err)
+	}
+
+	schemaBytes, err := json.Marshal(outputSchema.Schema)
+	if err != nil {
+		return fmt.Errorf("expected schema is not valid: %w", err)
+	}
+
+	schemaLoader := gojsonschema.NewStringLoader(string(schemaBytes))
+	jsonLoader := gojsonschema.NewGoLoader(jsonData)
+	result, err := gojsonschema.Validate(schemaLoader, jsonLoader)
+	if err != nil {
+		return fmt.Errorf("failed to validate expected schema: %w", err)
+	}
+
+	if !result.Valid() {
+		var errMsg string
+		for _, err := range result.Errors() {
+			errMsg += fmt.Sprintf("- %s\n", err)
```

Comment:
If `result.Errors` is a `[]error`, you can write
```
err := errors.Join(result.Errors()...)
```

Reply:
It is `[]gojsonschema.ResultError` unfortunately.